[FSDP] Add initial `summon_full_params(with_grads=True)`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #86122 [FSDP][2/N] Remove `_fsdp_wrapped_module.flat_param`
* #86117 [FSDP][1/N] Retire `FlattenParamsWrapper`
* **#85738 [FSDP] Add initial `summon_full_params(with_grads=True)`**
* #84911 [FSDP] Add `use_orig_params`

This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

Commit:
[FSDP] Add initial `summon_full_params(with_grads=True)`

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

[ghstack-poisoned]

rohan-varma commented on test/distributed/fsdp/test_fsdp_summon_full_params.py:
```diff
@@ -578,6 +581,100 @@ def test_named_parameters_buffers(self, prefix: str, recurse: bool):
+                "writeback": [False, True],
+                "offload_to_cpu": [False, True],
+                "sharding_strategy": [
+                    ShardingStrategy.FULL_SHARD,
```
also shard_grad_op?

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -2945,7 +2955,10 @@ def _summon_full_params(
         writeback: bool = True,
         rank0_only: bool = False,
         offload_to_cpu: bool = False,
+        with_grads: bool = False,
```
maybe add this to the docstring since it is a public facing API, but document the limited support and what happens when grads are None?

rohan-varma reviewed (approved):
LGTM overall, some mostly minor questions / comments.

awgu commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -846,6 +846,36 @@ def _free_low_precision_sharded_param(self):
+            flat_param._saved_grad_shard = flat_param.grad  # type: ignore[attr-defined]
+            sharded_grad = flat_param._saved_grad_shard  # type: ignore[attr-defined]
+        dist._all_gather_base(padded_unsharded_grad, sharded_grad, self.process_group)
+        unsharded_size = self.flat_param._unpadded_unsharded_size
```
Yes. Sorry that the names are confusing. Hopefully, they will make more sense with more familiarity. Unsharded implies unflattened* and sharded implies flattened.

*For `use_orig_params=False`, the unflatten happens in the `FlattenParamsWrapper.forward()`, so there is a small window between the the pre-forward and forward when the `flat_param` is unsharded but not unflattened. For `use_orig_params=True`, I minimized that window to be basically a single line within the same method. I will make the same change for `use_orig_params=False` in a follow-up PR.

awgu commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -1139,6 +1169,40 @@ def _use_unsharded_views(self, as_params: bool) -> None:
+            )
+            param = getattr(module, param_name)
+            param.grad = view
+        for i, (
```
This just makes sure that each shared parameter's `.grad` is also populated. Without this, then if we had a model like
```
lin = nn.Linear(5, 5)
lin.weight = lin.bias
```
then only one of `(lin.weight, lin.bias)` would get a `.grad` since only one of them is in accounted for in `_param_infos` / `_params` / etc.

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -846,6 +846,36 @@ def _free_low_precision_sharded_param(self):
+            self._check_sharded(flat_param.grad)
+            flat_param._saved_grad_shard = flat_param.grad  # type: ignore[attr-defined]
+            sharded_grad = flat_param._saved_grad_shard  # type: ignore[attr-defined]
+        dist._all_gather_base(padded_unsharded_grad, sharded_grad, self.process_group)
```
We discussed briefly offline, but there are two options:
1. (As in the PR currently) We only use a single all-gather collective per `FlatParameter`. In that case, if all ranks' sharded gradient is `None`, then the unsharded gradient is incorrectly `torch.zeros(unsharded_size)`. If only some ranks' sharded gradients are `None`, then the unsharded gradient zeros those corresponding elements.
2. We use a preceding all-reduce collective per `FlatParameter` to indicate if each rank's sharded gradient is `None` or not. This solves the problem from 1.

Since `summon_full_params(with_grads=True)` is meant for debugging, I can see the argument for pursuing 2. I can change this in a follow-up PR and add/adjust unit tests accordingly.

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

Commit:
Update on "[FSDP] Add initial `summon_full_params(with_grads=True)`"


This adds `summon_full_params(with_grads=True)` for `use_orig_params=True` and `offload_to_cpu=False`. Filling in the `use_orig_params=False` case requires some already-planned refactoring, and the `offload_to_cpu=True` case needs some additional work as well.

Adding this is helpful for debugging `use_orig_params=True` to make sure gradients are being updated correctly.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

zhaojuanmao commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -847,6 +847,37 @@ def _free_low_precision_sharded_param(self):
+            sharded_grad = torch.zeros_like(flat_param)  # type: ignore[attr-defined]
+        else:
+            self._check_sharded(flat_param.grad)
+            flat_param._saved_grad_shard = flat_param.grad  # type: ignore[attr-defined]
```
curious why we intentionally fill "flat_param._saved_grad_shard" here?

awgu replied:
We already use `_saved_grad_shard` for saving the sharded gradient during the backward pass. Here, I use the same variable, now for a different purpose: saving the sharded gradient during `summon_full_params(with_grads=True)`. This is just to avoid creating an entirely new variable also for saving the sharded gradient.

zhaojuanmao reviewed (commented):
Thanks a lot for adding this! It will improve debug ability a lot under summon_full_params mode for users
